src/goto-instrument/contracts/utils.h

```diff
@@ -190,6 +190,34 @@ class cfg_infot
     return locals.is_local(ident) || symbol.is_parameter;
   }
 
+  /// Returns true iff `expr` is a composite type access on a locally declared
+  /// or parameter symbol, without any dereference operations.
+  bool is_access_to_local_composite(const exprt &expr) const
```

Comment:
I believe typecasts and byte extract expressions are also possible left-hand sides in goto programs.